File: python_code/bin/binary_counter.py
import RPi.GPIO as GPIO
import time
import sys

# Check Python version
is_python2 = sys.version_info[0] == 2

# Setup
GPIO.setwarnings(False)
GPIO.setmode(GPIO.BOARD)

# Define the 5 LED pins using your specified pins
# From least significant bit (2^0) to most significant bit (2^4)
LED_PINS = [37, 35, 33, 31, 29]  # Your specified pins
# Note: Pin 39 is used for ground, not controlled by software

# Setup all pins as outputs
for pin in LED_PINS:
    GPIO.setup(pin, GPIO.OUT)
    GPIO.output(pin, False)  # Start with all LEDs off

def display_binary(number):
    """Display a number (0-31) in binary using 5 LEDs"""
    if number < 0 or number > 31:
        print("Number must be between 0 and 31")
        return
    
    binary = bin(number)[2:].zfill(5)  # Convert to binary and pad to 5 digits
    print("Decimal: {} | Binary: {}".format(number, binary))
    
    # First turn off all LEDs
    for pin in LED_PINS:
        GPIO.output(pin, False)
    
    # Turn on LEDs based on binary representation
    # Note: binary string is left-to-right (MSB to LSB)
    # but we want to map it to LEDs right-to-left (LSB to MSB)
    for i, bit in enumerate(reversed(binary)):
        if bit == '1':
            GPIO.output(LED_PINS[i], True)
    
    # Debug output
    led_states = []
    for pin in LED_PINS:
        led_states.append("ON" if GPIO.input(pin) else "OFF")
    print("LED states: {}".format(led_states))

def get_input(prompt):
    """Compatible input function for both Python 2 and Python 3"""
    if is_python2:
        return raw_input(prompt)
    else:
        return input(prompt)

try:
    # Interactive mode
    while True:
        print("\nBinary Counter (0-31)")
        print("Options:")
        print("1. Count automatically from 0 to 31")
        print("2. Enter a specific number")
        print("3. Quit")
        
        choice = get_input("Enter your choice (1-3): ").strip()
        
        if choice == '1':
            print("Counting from 0 to 31...")
            for i in range(32):
                display_binary(i)
                time.sleep(1)
        elif choice == '2':
            try:
                num = int(get_input("Enter a number (0-31): ").strip())
                if 0 <= num <= 31:
                    display_binary(num)
                else:
                    print("Number must be between 0 and 31")
            except ValueError:
                print("Please enter a valid number")
        elif choice == '3':
            print("Exiting...")
            break
        else:
            print("Invalid choice, please try again")

except KeyboardInterrupt:
    print("\nProgram stopped by user")
finally:
    # Turn off all LEDs and clean up
    for pin in LED_PINS:
        GPIO.output(pin, False)
    GPIO.cleanup()
    print("GPIO cleanup completed")

File: rust_code/pi_afr/src/bin/binary_counter.rs
use rppal::gpio::{Gpio, OutputPin};
use std::thread::sleep;
use std::time::Duration;
use std::error::Error;
use std::io::{self, Write};

// Define the 5 LED pins using BOARD pins 37, 35, 33, 31, 29 converted to BCM
// Note: rppal uses BCM pin numbering, not BOARD (physical) numbers
// BOARD pin 37 = BCM 26, BOARD pin 35 = BCM 19, BOARD pin 33 = BCM 13, 
// BOARD pin 31 = BCM 6, BOARD pin 29 = BCM 5
const LED_PINS: [u8; 5] = [26, 19, 13, 6, 5];  // BCM numbering for your pins

struct BinaryCounter {
    leds: Vec<OutputPin>,
}

impl BinaryCounter {
    fn new() -> Result<Self, Box<dyn Error>> {
        let gpio = Gpio::new()?;
        let mut leds = Vec::new();
        
        // Initialize all LED pins
        for &pin in &LED_PINS {
            let mut output_pin = gpio.get(pin)?.into_output();
            output_pin.set_low();  // Start with all LEDs off
            leds.push(output_pin);
        }
        
        Ok(BinaryCounter { leds })
    }
    
    fn display_binary(&mut self, number: u8) -> Result<(), Box<dyn Error>> {
        if number > 31 {
            println!("Number must be between 0 and 31");
            return Ok(());
        }
        
        // Convert to binary string
        let binary = format!("{:05b}", number);
        println!("Decimal: {} | Binary: {}", number, binary);
        
        // First turn off all LEDs
        for led in &mut self.leds {
            led.set_low();
        }
        
        // Turn on LEDs based on binary representation
        // Note: binary string is left-to-right (MSB to LSB)
        // but we want to map it to LEDs right-to-left (LSB to MSB)
        let binary_chars: Vec<char> = binary.chars().collect();
        for i in 0..binary_chars.len() {
            if binary_chars[binary_chars.len() - 1 - i] == '1' {
                self.leds[i].set_high();
            }
        }
        
        // Debug output
        let mut led_states = Vec::new();
        for i in 0..self.leds.len() {
            if binary_chars[binary_chars.len() - 1 - i] == '1' {
                led_states.push("ON");
            } else {
                led_states.push("OFF");
            }
        }
        println!("LED states: {:?}", led_states);
        
        Ok(())
    }
    
    fn reset(&mut self) {
        for led in &mut self.leds {
            led.set_low();
        }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let mut counter = BinaryCounter::new()?;
    
    println!("Binary Counter (0-31)");
    
    loop {
        println!("\nOptions:");
        println!("1. Count automatically from 0 to 31");
        println!("2. Enter a specific number");
        println!("3. Quit");
        
        print!("Enter your choice (1-3): ");
        io::stdout().flush()?;
        
        let mut choice = String::new();
        io::stdin().read_line(&mut choice)?;
        
        match choice.trim() {
            "1" => {
                println!("Counting from 0 to 31...");
                for i in 0..32 {
                    counter.display_binary(i)?;
                    sleep(Duration::from_secs(1));
                }
            },
            "2" => {
                print!("Enter a number (0-31): ");
                io::stdout().flush()?;
                
                let mut input = String::new();
                io::stdin().read_line(&mut input)?;
                
                match input.trim().parse::<u8>() {
                    Ok(num) if num <= 31 => {
                        counter.display_binary(num)?;
                    },
                    Ok(_) => println!("Number must be between 0 and 31"),
                    Err(_) => println!("Please enter a valid number"),
                }
            },
            "3" => {
                println!("Exiting...");
                break;
            },
            _ => println!("Invalid choice, please try again"),
        }
    }
    
    // Clean up before exiting
    counter.reset();
    println!("Program completed");
    
    Ok(())
}
